File: 3013-divide-an-array-into-subarrays-with-minimum-cost-ii/3013-divide-an-array-into-subarrays-with-minimum-cost-ii.cpp
class Solution {
public:
    long long minimumCost(vector<int>& nums, int k, int dist) {
        long long res = LLONG_MAX;
        int n = nums.size();
        int l = 2;
        int r = 2;
        multiset<int> window;
        multiset<int> smallest;
        long long windowSum = nums[l];
        smallest.insert(nums[l]);
        for(int i = 1; i < n - k + 2; i++)
        { 
            while(r - l + 1 < min(dist, n - i - 1))
            {
                r++;
                if(smallest.size() < k - 2)
                {
                    smallest.insert(nums[r]);
                    windowSum += nums[r];
                }
                else
                {
                    auto lastEle = next(smallest.end(), -1);
                    if(nums[r] < *lastEle)
                    {
                        windowSum += -*lastEle + nums[r];
                        window.insert(*lastEle);
                        smallest.erase(lastEle);
                        smallest.insert(nums[r]);
                    }
                    else
                    {
                        window.insert(nums[r]);
                    }
                }
            }
            res = min(res, nums[0] + nums[i] + windowSum);
            if(smallest.find(nums[l]) != smallest.end())
            {
                smallest.erase(smallest.find(nums[l]));
                windowSum -= nums[l];
                if(!window.empty())
                {
                    smallest.insert(*(window.begin()));
                    windowSum += *(window.begin());
                    window.erase(window.begin());
                }
            }
            else
            {
                window.erase(window.find(nums[l]));
            }
            l++;
        }
        return res;
    }
};

File: 3013-divide-an-array-into-subarrays-with-minimum-cost-ii/3013-divide-an-array-into-subarrays-with-minimum-cost-ii.py
class Solution:
    def minimumCost(self, nums: List[int], k: int, dist: int) -> int:
        res = float('inf')
        l = 2
        r = 2
        window = SortedList()
        smallest = SortedList([nums[l]])
        windowSum = nums[l]
        for i in range(1, len(nums) - k + 2):
            print(i)
            while r - l + 1 < min(dist, len(nums) - i - 1):
                r += 1
                print(r - l + 1)
                if len(smallest) < k - 2:
                    smallest.add(nums[r])
                    windowSum += nums[r]
                else:
                    if nums[r] < smallest[-1]:
                        windowSum += -smallest[-1] + nums[r]
                        window.add(smallest[-1])
                        smallest.pop()
                        smallest.add(nums[r])
                    else:
                        window.add(nums[r])
            res = min(res, nums[0] + nums[i] + windowSum)
            if nums[l] in smallest:
                smallest.remove(nums[l])
                windowSum -= nums[l]
                if len(window) > 0:
                    smallest.add(window[0])
                    windowSum += window[0]
                    window.pop(0)
            else:
                window.remove(nums[l])
            l += 1
        return res